solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+#![allow(dead_code)]
+
+use super::single_pair_settlement::{AmmSwapExactTokensForTokens, SinglePairSettlement};
+use crate::settlement::Trade;
+use anyhow::{anyhow, Result};
+use model::order::{OrderCreation, OrderKind};
+use num::{bigint::Sign, BigInt};
+use std::collections::HashMap;
+use web3::types::{Address, U256};
+
+#[derive(Default)]
+struct TokenContext {
+    address: Address,
+    reserve: U256,
+    buy_volume: U256,
+    sell_volume: U256,
+}
+
+impl TokenContext {
+    pub fn is_excess(&self, other: &TokenContext) -> bool {
+        u256_to_bigint(&self.reserve)
+            * (u256_to_bigint(&other.sell_volume) - u256_to_bigint(&other.buy_volume))
+            < u256_to_bigint(&other.reserve)
+                * (u256_to_bigint(&self.sell_volume) - u256_to_bigint(&self.buy_volume))
+    }
+}
+
+/**
+ * Computes a settlement using orders of a single pair and the direct AMM between those tokens.
+ * Panics orders are not already filtered for a specific token pair, or the reserve information for that
+ * pair is not available.
+ */
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    reserves: &HashMap<Address, U256>,
+) -> SinglePairSettlement {
+    let (context_a, context_b) = split_into_contexts(orders.clone(), reserves);
+    let (shortage, excess) = if context_a.is_excess(&context_b) {
+        (context_b, context_a)
+    } else {
+        (context_a, context_b)
+    };
+    let uniswap_out = compute_uniswap_out(&shortage, &excess);
+    let uniswap_in = compute_uniswap_in(uniswap_out, &shortage, &excess);
+    let interaction = if !uniswap_out.is_zero() {
+        Some(AmmSwapExactTokensForTokens {
+            amount_in: uniswap_in,
+            amount_out_min: uniswap_out,
+            token_in: excess.address,
+            token_out: shortage.address,
+        })
+    } else {
+        // TODO(fleupold) set correct clearing prices when supply/demand already match current uniswap price
+        None
+    };
+    // TODO(fleupold) check that all orders comply with the computed price. Otherwise, remove the least favorable excess order and try again.
+    SinglePairSettlement {
+        clearing_prices: maplit::hashmap! {
+            shortage.address => uniswap_in,
+            excess.address => uniswap_out,
```

Comment:
Can't this theoretically have 0-prices? Is that what you meant by the `TODO` on line 53?

Reply:
Yes.